### PR TITLE
PD-1034: Makes icons accessible

### PR DESCRIPTION
**Pull request by brookescarlett**
## ✍️ Proposed changes

(Previously agreed upon mechanism to implement accessible icons detailed in ticket)
- Icons now use `aria-label` rather than a `title` prop to serve as an accessible label
- Ensures that `role` can be overwritten, such that Icons can be hidden from screenreaders
- Additional accessibility rules fork on role, which can be either `presentation` or `img`
- Updates downstream tests which were selecting icons by title

🎟 _Jira ticket:_ [PD-1034](https://jira.mongodb.org/browse/PD-1034)

## 🛠 Types of changes

- [ ] Tooling (updates to workspace config or internal tooling)
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## ✅ Checklist

- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] I have run `yarn changeset` and documented my changes

- [ ] I have added my new package to the global tsconfig
- [ ] I have added my new package to the Table of Contents on the global README

### CODE REVIEW

File: packages/icon/src/glyphCommon.ts [hunk 1]
```diff
@@ -14,15 +14,34 @@ export const sizeMap: Record<Size, number> = {
...
+      } else if (accessibleParams.ariaLabel) {
+        return { 'aria-label': accessibleParams.ariaLabel };
+      } else if (accessibleParams.title) {
+        return { title: accessibleParams.title };
+      } else {
+        return { 'aria-label': getGlyphLabel(glyphName) };
+      }
+
+    case 'presentation':
+      return { 'aria-hidden': true };
```
[the remark this answers is not in the document]
Comment (brookescarlett) [reply packages/icon/src/glyphCommon.ts:41]: If it's role="presentation" you're indicating that you don't want this to be accessible to a screenreader, so the other aria params are no longer relevant. I can document this better though, if you think that's needed.

File: packages/select/src/Select.spec.tsx [hunk 2]
```diff
@@ -316,9 +316,11 @@ describe('packages/select', () => {
           expect(getByRole('listbox')).toBeVisible();
         });
 
-        getAllByTitle('Beaker Icon').forEach(element =>
-          expect(element.closest('svg')).toBeVisible(),
-        );
+        getAllByRole('img')
+          .filter(
+            element => element.getAttribute('aria-label') === 'Beaker Icon',
```
[the remark this answers is not in the document]
Comment (brookescarlett) [reply packages/select/src/Select.spec.tsx:321]: Wasn't aware that `getByLabelText` grabs aria-label. I thought that was strictly the label tag. If that works, I can do that.

File: packages/icon/src/glyphCommon.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (lazytype) [reply packages/icon/src/glyphCommon.ts:41]: Documenting is good, also enforcing at the type level would be great

File: packages/icon/src/glyphCommon.ts [hunk 3]
```diff
@@ -14,15 +14,34 @@ export const sizeMap: Record<Size, number> = {
...
+  ariaLabelledby?: string;
+  ariaLabel?: string;
+}
 
-  if (title == null || title === true) {
-    return `${name.replace(/([a-z])([A-Z])/g, '$1 $2')} Icon`;
+export function generateAccessibleProps(
+  role: 'img' | 'presentation',
+  glyphName: string,
+  accessibleParams: AccessibleFunctionParams,
```
[the remark this answers is not in the document]
Comment (brookescarlett) [reply packages/icon/src/glyphCommon.ts:26]: 👍

File: packages/select/src/Select.spec.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (brookescarlett) [reply packages/select/src/Select.spec.tsx:321]: Ah you're right about this one. I wish I knew before I updated all of the test suites 😅

[the remark this answers is not in the document]
Comment (lazytype) [reply packages/select/src/Select.spec.tsx:321]: I believe it even works with `aria-labelledby`

[the remark this answers is not in the document]
Comment (brookescarlett) [reply packages/select/src/Select.spec.tsx:321]: Indeed you're correct

File: packages/icon/src/glyphCommon.ts [hunk 4]
```diff
@@ -14,15 +14,34 @@ export const sizeMap: Record<Size, number> = {
...
+) {
```
[the remark this answers is not in the document]
Comment (brookescarlett) [reply packages/icon/src/glyphCommon.ts:27]: 👍

File: packages/icon/src/glyphCommon.ts [hunk 5]
```diff
@@ -14,15 +14,34 @@ export const sizeMap: Record<Size, number> = {
...
+  switch (role) {
+    case 'img':
+      if (accessibleParams.ariaLabelledby) {
+        return { 'aria-labelledby': accessibleParams.ariaLabelledby };
+      } else if (accessibleParams.ariaLabel) {
+        return { 'aria-label': accessibleParams.ariaLabel };
```
[the remark this answers is not in the document]
Comment (brookescarlett) [reply packages/icon/src/glyphCommon.ts:33]: I'm seeing this in the accessibility specs: `As required by the text alternative computation, user agents give precedence to aria-labelledby over aria-label when computing the accessible name property.` 

I believe that multiple properties don't contribute to the accessible name of an element. They've defined an order of precedence that gets followed here: https://www.w3.org/TR/wai-aria/#aria-labelledby

[the remark this answers is not in the document]
Comment (lazytype) [reply packages/icon/src/glyphCommon.ts]: if that's the case then why don't we pass both props through? the browser will handle ignoring the prop. alternatively we prevent providing both props by typing `LGGlyph.ComponentProps` to do so

File: packages/select/src/Select.spec.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (brookescarlett) [reply packages/select/src/Select.spec.tsx]: 👍

File: packages/icon/src/glyphCommon.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (brookescarlett) [reply packages/icon/src/glyphCommon.ts]: Does something like this make more sense to you: 

```
export function generateAccessibleProps(
  role: 'img' | 'presentation',
  glyphName: string,
  { ariaLabel, ariaLabelledby, title }: AccessibleFunctionParams,
): AccessibleFunctionReturnType {
  switch (role) {
    case 'img':
      if (!ariaLabel && !ariaLabelledby && !title) {
        return { 'aria-label': getGlyphLabel(glyphName) };
      }

      return {
        ['aria-labelledby']: ariaLabelledby,
        ['aria-label']: ariaLabel,
        title,
      };

    case 'presentation':
      return { 'aria-hidden': true };
  }
}
```

[the remark this answers is not in the document]
Comment (lazytype) [reply packages/icon/src/glyphCommon.ts]: 👍

File: packages/typography/src/Link.spec.tsx [hunk 6]
```diff
@@ -56,7 +56,7 @@ describe('packages/typography', () => {
         });
         const anchor = screen.getByText('Link');
         const icon = screen.queryByLabelText('Arrow Right Icon');
-        expect(icon).not.toBeVisible();
+        // expect(icon).not.toBeVisible();
```
Comment (brookescarlett) [inline packages/typography/src/Link.spec.tsx:59]: So this is a faulty test in general. We're testing for a lack of visibility by default and then not checking that it is actually visible when we hover - only that the Icon is in the document (which doesn't prove anything). When I try and update the test to be icon.toBeVisible() that fails too. I can inspect the css styles in the browser-rendered version of the test suite and I see that the styles are correctly applied -- the only thing that I can suspect that's preventing React Testing library from evaluating the Icon as visible is that it's not a straightforward css rule that transitions the opacity, and maybe that's causing the issue. Have been struggling with this for a while, and I'm not too sure what to do here.

Comment (brookescarlett) [reply packages/typography/src/Link.spec.tsx]: I'm leaning towards removing the test entirely and filing a ticket to create a better test here. I'm almost positive that this Icon updates did not break the test, but rather exposed that the test never worked as expected in the first place. What do you think @lazytype

Comment (lazytype) [reply packages/typography/src/Link.spec.tsx]: > the only thing that I can suspect that's preventing React Testing library from evaluating the Icon as visible is that it's not a straightforward css rule that transitions the opacity

for transitions I've had some success by waiting for them to finish, e.g. `await waitFor(() => expect(icon).not.toBeVisible());`  (`waitForElementToBeRemoved` if the element is no longer in the document at all). if this approach still doesn't work then I'm ok with creating a ticket and linking it in the file

Comment (brookescarlett) [reply packages/typography/src/Link.spec.tsx]: Still doesn't work, even though I'm seeing this rule applied: 

`[data-leafygreen-ui="anchor-container"]:hover .leafygreen-ui-vo4x0d{opacity:1;-webkit-transform:translate3d(3px,0,0);-ms-transform:translate3d(3px,0,0);transform:translate3d(3px,0,0);}`

Comment (brookescarlett) [reply packages/typography/src/Link.spec.tsx]: Rming this test: 

```
test('and the "arrowAppearance" prop is set to "hover"', async () => {
        renderLink({
          href: 'http://localhost:9001',
          arrowAppearance: 'hover',
        });
        const anchor = screen.getByText('Link').parentNode;
        const icon = screen.queryByLabelText('Arrow Right Icon');
        expect(icon).not.toBeVisible();

        fireEvent.mouseEnter(anchor);
        await waitFor(() => expect(icon).toBeVisible())
      });
```

Comment (brookescarlett) [reply packages/typography/src/Link.spec.tsx]: https://jira.mongodb.org/browse/PD-1090

File: packages/typography/src/Link.spec.tsx [hunk 7]
```diff
@@ -52,53 +49,34 @@ describe('packages/typography', () => {
         });
       });
 
-      test('and the "arrowAppearance" prop is set to "hover"', () => {
```
Comment (lazytype) [inline packages/typography/src/Link.spec.tsx:55]: I'd prefer adding a `test.todo(` and add a commenting linking to the ticket

File: packages/icon/src/glyphCommon.ts [hunk 8]
```diff
@@ -14,15 +14,41 @@ export const sizeMap: Record<Size, number> = {
...
+  ariaLabelledby?: string;
+  ariaLabel?: string;
+}
+
+interface AccessibleFunctionReturnType {
+  'aria-labelledby'?: string;
+  'aria-label'?: string;
+  title?: string | null;
+  'aria-hidden'?: boolean;
+}
```
Comment (lazytype) [inline packages/icon/src/glyphCommon.ts:28]: this can now be a union of the two possible structures

File: packages/icon/src/glyphCommon.ts [hunk 9]
```diff
@@ -14,15 +14,44 @@ export const sizeMap: Record<Size, number> = {
...
-export function getGlyphTitle(name: string, title?: string | boolean | null) {
-  if (title === false) {
-    // If title is null, we unset the title entirely, otherwise we generate one.
-    return null;
-  }
+interface AccessibleFunctionParams {
+  title?: string | null;
+  ariaLabelledby?: string;
+  ariaLabel?: string;
+}
```
Comment (lazytype) [inline packages/icon/src/glyphCommon.ts:21]: delete

File: packages/icon/src/glyphCommon.ts [hunk 10]
```diff
@@ -14,15 +14,44 @@ export const sizeMap: Record<Size, number> = {
...
-  if (title == null || title === true) {
-    return `${name.replace(/([a-z])([A-Z])/g, '$1 $2')} Icon`;
+type AccessibleFunctionReturnType =
+  | AccessibleFunctionParams
+  | { 'aria-hidden': true };
+
+export function generateAccessibleProps(
+  role: 'img' | 'presentation',
+  glyphName: string,
+  { ariaLabel, ariaLabelledby, title }: AccessibleFunctionParams,
```
Comment (lazytype) [inline packages/icon/src/glyphCommon.ts:36]:
```suggestion
  { 'aria-label': ariaLabel, 'aria-labelledby': ariaLabelledby, title }: AccessibleFunctionParams,
```

File: packages/icon/src/glyphCommon.ts (hunk 9 above)
Comment (brookescarlett) [reply packages/icon/src/glyphCommon.ts:21]: can't delete bc we use camel case spelling for the inputs and then kebab case for the outputs, but renamed

File: packages/icon/src/glyphCommon.ts (hunk 10 above)
Comment (brookescarlett) [reply packages/icon/src/glyphCommon.ts]: let me know if you prefer this approach

**Review by lazytype — APPROVED** [review]

**Pull request merged**